### ISSUE TITLE
version attribute not required

### ISSUE THREAD

**Issue by sydb** [xproc/3.0-grammar#20]
The xproc30 schema does not require `/p:declare-step/@version`, but

1. morgana does; and
1. [section 13 of the spec ](https://spec.xproc.org/master/head/xproc/#versioning-considerations) says it is required (and that its value must be "3.0" if you want the thing to work).

**Comment by xml-project** [comment]
Well, yes and no. The version attribute is only required on what the spec calls the pipeline document, i.e. the outermost p:declare / p:library. Every p:declare-step inside the pipeline document inherits the version attribute's value, if it is not specified.

**Comment by sydb** [comment]
Agreed (hence the initial slash in the path above).

Thus I am suggesting something along the lines of dividing each of the patterns “DeclareStep”, “DeclarePipleStep”, and “DeclareAtomicStep” into two, where the one for use on `start` would depend on “version.3.0.attr” (or some such) which would be declared as "3.0" and be required.

I realize this is a lot of work for not a whole lot of gain, but in the sense that the purpose of a schema is to catch errors early in the process, seems like a good idea.

If y’all used Schematron, too, a simple
```xml
    <sch:rule context="/*">
      <sch:assert test="normalize-space(@version) eq '3.0'">
        The outermost <sch:value-of select="name(.)"/> should be
        declared as being XProc version 3 (with a version="3.0")
      </sch:assert>
    </sch:rule>
```
would probably be a good compromise.

P.S. In principle I do not mind doing the work and submitting a PR, but I do not understand which bits in the repo are source and which are generated. (And had never heard of gradle until this. :-) So if someone wants to tell me which file to edit and how to test …

**Comment by gimsieke** [comment]
The relevant source file is https://github.com/xproc/3.0-grammar/blob/master/src/main/schema/core30.rnc
If you file a PR, a CI test will be launched, so you don’t need to fiddle with Gradle.

Since the code snippets in https://spec.xproc.org/master/head/xproc/#declare-pipelines and https://spec.xproc.org/master/head/xproc/#p.library are generated by referencing their pattern names, you should refrain from renaming the patterns that contain the optional version attribute. 
But you certainly can add the Schematron rule instead. In a future oXygen XProc framework, for example, embedded Schematron rules can be applied by default.
Since core30.rnc also specifies other document element names such as `/c:param-set`, you should use `context="/p:declare-step | /p:library`. (And I’d convert the RNC to RNG first, then insert the Schematron, and then convert it back and paste the generated RNC Schematron into core30.rnc.)

**Comment by sydb** [comment]
Thanks @gimsieke! I have submitted a PR (#21) that makes the changes to the RELAX NG. I don’t have any capability to really test these changes locally, but
 a) I did check the validity of core3.rnc — after adding a dummy definition of “StandardStep”, it is at least a valid RELAX NG schema;
 b) it looks like Travis has run a check, and it passed that.

I heeded your warning and did not change the existing pattern names.

And I agree with you completely about converting to XML syntax before adding Schematron. I drastically prefer the compact syntax until the moment something has to be added in another namespace. Then suddenly I want the XML syntax. :-)
